File: server/src/protocol.ts
// Wire protocol contract — mirror of docs/PROTOCOL.md and Editor/Core/Protocol.cs.
// Any change lands in all three or not at all.

export const PROTOCOL_V = 1;
export const MAX_FRAME_BYTES = 64 * 1024 * 1024;
export const MAX_IN_FLIGHT = 32;
export const HELLO_TIMEOUT_MS = 5_000;
export const PORT_BASE = 47700;
export const PORT_SLOTS = 64;
export const PORT_PROBE_STEPS = 8;

export type FrameType =
  | "hello"
  | "welcome"
  | "req"
  | "res"
  | "progress"
  | "event"
  | "cancel"
  | "ping"
  | "pong"
  | "bye";

export interface Envelope<T = unknown> {
  v?: number; // hello/welcome only
  id: string; // uuid; res/progress echo the req id
  type: FrameType;
  payload: T;
}

export interface HelloPayload {
  v: { min: number; max: number };
  client: {
    name: string;
    version: string;
    pid: number;
    sessionId: string;
    /** Registry token (v2.1+): echoed from the registry entry when present. */
    token?: string;
  };
  features: string[];
}

export interface WelcomePayload {
  v: number;
  plugin: { version: string };
  unity: { version: string; projectPath: string; projectName: string };
  editor: { sessionId: string; pid: number; domainReloadCount: number };
  eval: { engine: "csc" | "codedom" | "none" };
  lease: { holder?: string };
  features: string[];
}

export interface ReqPayload {
  method: string;
  params: unknown;
  timeoutMs?: number;
  leaseHint?: string;
}

export interface ResPayload {
  ok: boolean;
  result?: unknown;
  error?: ErrorObj;
}

export interface ProgressPayload {
  pct?: number;
  message?: string;
  phase?: string;
  seq: number;
}

export interface EventPayload {
  kind: EventKind;
  data: unknown;
}

export type EventKind =
  | "log"
  | "compile.started"
  | "compile.finished"
  | "reload.imminent"
  | "playmode.changed"
  | "job.progress"
  | "job.terminal"
  | "lease.lost";

export interface CancelPayload {
  targetId: string;
}

export interface ByePayload {
  reason: "domain_reload" | "quit" | "shutdown";
  resumeHintMs?: number;
}

export interface Diagnostic {
  file: string;
  line: number;
  col: number;
  severity: "error" | "warning";
  csCode: string;
  text: string;
}

export interface ErrorObj {
  code: ErrorCode;
  message: string;
  retryable: boolean;
  detail?: unknown;
  unityStack?: string;
  consoleErrors?: string[];
  diagnostics?: Diagnostic[];
}

export type ErrorCode =
  // plugin-side
  | "PARSE_ERROR"
  | "PROTOCOL_ERROR"
  | "VERSION_UNSUPPORTED"
  | "AUTH_REQUIRED"
  | "HELLO_TIMEOUT"
  | "METHOD_NOT_FOUND"
  | "INVALID_PARAMS"
  | "HANDLER_EXCEPTION"
  | "TIMEOUT"
  | "CANCELLED"
  | "DOMAIN_RELOAD"
  | "BUSY_MODAL"
  | "LEASE_HELD"
  | "LEASE_LOST"
  | "JOB_NOT_FOUND"
  | "JOB_NOT_RESUMABLE"
  | "EVAL_COMPILE_ERROR"
  | "EVAL_RUNTIME_ERROR"
  | "EVAL_ENGINE_UNAVAILABLE"
  // server-synthesized
  | "UNITY_UNREACHABLE"
  | "PROJECT_NOT_FOUND"
  | "PROJECT_AMBIGUOUS"
  | "RECONNECT_TIMEOUT"
  // VCC/VPM layer (v2.4.0, server-local - no plugin involved)
  | "VRC_GET_NOT_FOUND"
  | "VRC_GET_FAILED";

export const RETRYABLE_CODES: ReadonlySet<ErrorCode> = new Set([
  "DOMAIN_RELOAD",
  "BUSY_MODAL",
]);

export interface RegistryEntry {
  schemaVersion: number;
  port: number;
  projectPath: string;
  projectName: string;
  pid: number;
  unityVersion: string;
  pluginVersion: string;
  protocolV: number;
  startedAt: string;
  /** Connection token (v2.1+): clients echo it in hello.client.token. */
  token?: string;
}

export interface JobRecord {
  jobId: string;
  method: string;
  state: "pending" | "running" | "completed" | "failed" | "cancelled";
  phase?: string;
  pct?: number;
  message?: string;
  ownerSessionId: string;
  startedAt: string;
  updatedAt: string;
  result?: unknown;
  error?: ErrorObj;
  reloadCount: number;
}

/** fnv1a32 over the normalized project path; used for port slot + registry filename. */
export function fnv1a32(input: string): number {
  let h = 0x811c9dc5;
  for (let i = 0; i < input.length; i++) {
    h ^= input.charCodeAt(i);
    h = Math.imul(h, 0x01000193) >>> 0;
  }
  return h >>> 0;
}

/** Absolute path → forward slashes, no trailing slash, lower-case (Windows). */
export function normalizeProjectPath(p: string): string {
  let s = p.replace(/\\/g, "/");
  while (s.length > 1 && s.endsWith("/")) s = s.slice(0, -1);
  return s.toLowerCase();
}

export function preferredPort(projectPath: string): number {
  return PORT_BASE + (fnv1a32(normalizeProjectPath(projectPath)) % PORT_SLOTS);
}

export function registryFileName(projectPath: string): string {
  return fnv1a32(normalizeProjectPath(projectPath)).toString(16).padStart(8, "0") + ".json";
}

File: package/com.tunasync.unity-mcp/Editor/Core/Protocol.cs
// Wire protocol contract - mirror of docs/PROTOCOL.md and server/src/protocol.ts.
// Any change lands in all three or not at all.
using System;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using Newtonsoft.Json.Serialization;

namespace TunaSync.UnityMCP.Editor
{
    public static class Protocol
    {
        public const int Version = 1;
        public const int MaxFrameBytes = 64 * 1024 * 1024;
        public const int MaxInFlight = 32;
        public const int HelloTimeoutMs = 5000;
        public const int PortBase = 47700;
        public const int PortSlots = 64;
        public const int PortProbeSteps = 8;
        public const int RegistrySchemaVersion = 1;

        public static readonly JsonSerializerSettings JsonSettings = new JsonSerializerSettings
        {
            ContractResolver = new DefaultContractResolver
            {
                NamingStrategy = new CamelCaseNamingStrategy()
            },
            NullValueHandling = NullValueHandling.Ignore,
            Formatting = Formatting.None,
        };

        public static string Serialize(object o) => JsonConvert.SerializeObject(o, JsonSettings);

        /// <summary>fnv1a32 over the normalized project path (port slot + registry filename).</summary>
        public static uint Fnv1a32(string input)
        {
            uint h = 0x811c9dc5;
            foreach (char c in input)
            {
                h ^= c;
                h *= 0x01000193;
            }
            return h;
        }

        /// <summary>Absolute path -> forward slashes, no trailing slash, lower-case.</summary>
        public static string NormalizeProjectPath(string p)
        {
            string s = p.Replace('\\', '/');
            while (s.Length > 1 && s.EndsWith("/")) s = s.Substring(0, s.Length - 1);
            return s.ToLowerInvariant();
        }

        public static int PreferredPort(string projectPath)
            => PortBase + (int)(Fnv1a32(NormalizeProjectPath(projectPath)) % PortSlots);

        public static string RegistryFileName(string projectPath)
            => Fnv1a32(NormalizeProjectPath(projectPath)).ToString("x8") + ".json";
    }

    public static class FrameType
    {
        public const string Hello = "hello";
        public const string Welcome = "welcome";
        public const string Req = "req";
        public const string Res = "res";
        public const string Progress = "progress";
        public const string Event = "event";
        public const string Cancel = "cancel";
        public const string Ping = "ping";
        public const string Pong = "pong";
        public const string Bye = "bye";
    }

    public static class ErrorCodes
    {
        public const string ParseError = "PARSE_ERROR";
        public const string ProtocolError = "PROTOCOL_ERROR";
        public const string VersionUnsupported = "VERSION_UNSUPPORTED";
        public const string HelloTimeout = "HELLO_TIMEOUT";
        public const string MethodNotFound = "METHOD_NOT_FOUND";
        public const string InvalidParams = "INVALID_PARAMS";
        public const string HandlerException = "HANDLER_EXCEPTION";
        public const string Timeout = "TIMEOUT";
        public const string Cancelled = "CANCELLED";
        public const string DomainReload = "DOMAIN_RELOAD";   // retryable
        public const string BusyModal = "BUSY_MODAL";         // retryable
        public const string LeaseHeld = "LEASE_HELD";
        public const string LeaseLost = "LEASE_LOST";
        public const string JobNotFound = "JOB_NOT_FOUND";
        public const string JobNotResumable = "JOB_NOT_RESUMABLE";
        public const string EvalCompileError = "EVAL_COMPILE_ERROR";
        public const string EvalRuntimeError = "EVAL_RUNTIME_ERROR";
        public const string EvalEngineUnavailable = "EVAL_ENGINE_UNAVAILABLE";
        public const string AuthRequired = "AUTH_REQUIRED";
        public const string PlayModeActive = "PLAY_MODE_ACTIVE";
    }

    public static class EventKind
    {
        public const string Log = "log";
        public const string CompileStarted = "compile.started";
        public const string CompileFinished = "compile.finished";
        public const string ReloadImminent = "reload.imminent";
        public const string PlaymodeChanged = "playmode.changed";
        public const string JobProgress = "job.progress";
        public const string JobTerminal = "job.terminal";
        public const string LeaseLost = "lease.lost";
    }

    [Serializable]
    public class Envelope
    {
        [JsonProperty("v", NullValueHandling = NullValueHandling.Ignore)]
        public int? V;
        [JsonProperty("id")]
        public string Id;
        [JsonProperty("type")]
        public string Type;
        [JsonProperty("payload")]
        public JObject Payload;
    }

    public class ErrorObj
    {
        [JsonProperty("code")] public string Code;
        [JsonProperty("message")] public string Message;
        [JsonProperty("retryable")] public bool Retryable;
        [JsonProperty("detail", NullValueHandling = NullValueHandling.Ignore)] public object Detail;
        [JsonProperty("unityStack", NullValueHandling = NullValueHandling.Ignore)] public string UnityStack;
        [JsonProperty("consoleErrors", NullValueHandling = NullValueHandling.Ignore)] public string[] ConsoleErrors;
        [JsonProperty("diagnostics", NullValueHandling = NullValueHandling.Ignore)] public Diagnostic[] Diagnostics;

        public static ErrorObj Make(string code, string message, bool retryable = false)
            => new ErrorObj { Code = code, Message = message, Retryable = retryable };
    }

    public class Diagnostic
    {
        [JsonProperty("file")] public string File;
        [JsonProperty("line")] public int Line;
        [JsonProperty("col")] public int Col;
        [JsonProperty("severity")] public string Severity;
        [JsonProperty("csCode")] public string CsCode;
        [JsonProperty("text")] public string Text;
    }

    public class RegistryEntry
    {
        [JsonProperty("schemaVersion")] public int SchemaVersion = Protocol.RegistrySchemaVersion;
        [JsonProperty("port")] public int Port;
        [JsonProperty("projectPath")] public string ProjectPath;
        [JsonProperty("projectName")] public string ProjectName;
        [JsonProperty("pid")] public int Pid;
        [JsonProperty("unityVersion")] public string UnityVersion;
        [JsonProperty("pluginVersion")] public string PluginVersion;
        [JsonProperty("protocolV")] public int ProtocolV = Protocol.Version;
        [JsonProperty("startedAt")] public string StartedAt;
        // Same-user auth: random per-session token; clients read it from this
        // file (user-ACL'd %LOCALAPPDATA%) and echo it in hello.client.token.
        [JsonProperty("token", NullValueHandling = NullValueHandling.Ignore)] public string Token;
    }
}
